File: backend/app/models/product.py
"""
Product model representing inventory items.
"""

from typing import Optional, Dict, Any
from datetime import datetime
from decimal import Decimal
from pydantic import BaseModel, Field, validator


class Product(BaseModel):
    """
    Product model representing a store inventory item.

    Attributes:
        id: Unique product identifier (UUID)
        user_id: Store owner UUID
        name: Product name
        description: Product description
        price: Product price
        category: Product category
        sku: Stock keeping unit
        image_url: URL to product image
        stock_quantity: Available stock quantity
        is_active: Product visibility status
        metadata: Additional product data as JSON
        created_at: Product creation timestamp
        updated_at: Last update timestamp
    """

    id: Optional[str] = Field(None, description="Product UUID")
    user_id: Optional[str] = Field(None, description="Store owner UUID")
    name: str = Field(..., max_length=255, description="Product name")
    description: Optional[str] = Field(None, description="Product description")
    price: Decimal = Field(..., gt=0, description="Product price")
    category: Optional[str] = Field(
        None, max_length=100, description="Product category"
    )
    sku: Optional[str] = Field(None, max_length=100, description="Stock keeping unit")
    image_url: Optional[str] = Field(
        None, max_length=500, description="Product image URL"
    )
    stock_quantity: int = Field(default=0, ge=0, description="Stock quantity")
    is_active: bool = Field(default=True, description="Product active status")
    metadata: Dict[str, Any] = Field(
        default_factory=dict, description="Additional data"
    )
    created_at: Optional[datetime] = Field(None, description="Creation timestamp")
    updated_at: Optional[datetime] = Field(None, description="Last update timestamp")

    @validator("price")
    def validate_price(cls, v):
        """Validate price is positive"""
        if v <= 0:
            raise ValueError("Price must be greater than 0")
        return v

    @validator("sku")
    def validate_sku(cls, v):
        """Validate SKU format (alphanumeric, hyphens, underscores)"""
        if v is not None and v.strip():
            import re

            if not re.match(r"^[a-zA-Z0-9-_]+$", v):
                raise ValueError(
                    "SKU can only contain letters, numbers, hyphens, and underscores"
                )
        return v

    @validator("image_url")
    def validate_image_url(cls, v):
        """Validate image URL starts with http/https"""
        if v is not None and v.strip():
            import re

            if not re.match(r"^https?://", v):
                raise ValueError("Image URL must start with http:// or https://")
        return v

    class Config:
        """Pydantic configuration"""

        json_schema_extra = {
            "example": {
                "id": "123e4567-e89b-12d3-a456-426614174000",
                "user_id": "123e4567-e89b-12d3-a456-426614174001",
                "name": "Wireless Headphones",
                "description": "High-quality wireless headphones with noise cancellation",
                "price": 129.99,
                "category": "Electronics",
                "sku": "WH-001",
                "image_url": "https://example.com/images/headphones.jpg",
                "stock_quantity": 50,
                "is_active": True,
            }
        }

    def to_dict(self) -> Dict[str, Any]:
        """
        Convert model to dictionary.

        Returns:
            Dictionary representation
        """
        data = self.model_dump()
        # Convert Decimal to float for JSON serialization
        if "price" in data and isinstance(data["price"], Decimal):
            data["price"] = float(data["price"])
        return data

    def to_card_dict(self) -> Dict[str, Any]:
        """
        Convert to card format for chat display.

        Returns:
            Dictionary with card-friendly fields
        """
        return {
            "id": self.id,
            "name": self.name,
            "description": self.description[:200] if self.description else None,
            "price": float(self.price),
            "category": self.category,
            "image_url": self.image_url,
        }

    def get_search_text(self) -> str:
        """
        Get combined search text for full-text search.

        Returns:
            Combined name and description text
        """
        text = self.name
        if self.description:
            text += " " + self.description
        if self.category:
            text += " " + self.category
        return text


class ProductCreate(BaseModel):
    """Schema for creating a new product"""

    name: str = Field(..., max_length=255)
    description: Optional[str] = None
    price: Decimal = Field(..., gt=0)
    category: Optional[str] = Field(None, max_length=100)
    sku: Optional[str] = Field(None, max_length=100)
    image_url: Optional[str] = Field(None, max_length=500)
    stock_quantity: int = Field(default=0, ge=0)
    metadata: Dict[str, Any] = Field(default_factory=dict)


class ProductUpdate(BaseModel):
    """Schema for updating a product"""

    name: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = None
    price: Optional[Decimal] = Field(None, gt=0)
    category: Optional[str] = Field(None, max_length=100)
    sku: Optional[str] = Field(None, max_length=100)
    image_url: Optional[str] = Field(None, max_length=500)
    stock_quantity: Optional[int] = Field(None, ge=0)
    is_active: Optional[bool] = None
    metadata: Optional[Dict[str, Any]] = None


__all__ = ["Product", "ProductCreate", "ProductUpdate"]

File: flutter_app/lib/models/product.dart
/// Product model representing inventory items
class Product {
  final String id;
  final String userId;
  final String name;
  final String? description;
  final double price;
  final String? category;
  final String? sku;
  final String? imageUrl;
  final int stockQuantity;
  final bool isActive;
  final Map<String, dynamic>? metadata;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Product({
    required this.id,
    required this.userId,
    required this.name,
    this.description,
    required this.price,
    this.category,
    this.sku,
    this.imageUrl,
    this.stockQuantity = 0,
    this.isActive = true,
    this.metadata,
    this.createdAt,
    this.updatedAt,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      price: (json['price'] as num).toDouble(),
      category: json['category'] as String?,
      sku: json['sku'] as String?,
      imageUrl: json['image_url'] as String?,
      stockQuantity: json['stock_quantity'] as int? ?? 0,
      isActive: json['is_active'] as bool? ?? true,
      metadata: json['metadata'] as Map<String, dynamic>?,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'name': name,
      'description': description,
      'price': price,
      'category': category,
      'sku': sku,
      'image_url': imageUrl,
      'stock_quantity': stockQuantity,
      'is_active': isActive,
      'metadata': metadata,
      'created_at': createdAt?.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }

  /// Convert to card format for chat display
  Map<String, dynamic> toCardJson() {
    return {
      'id': id,
      'name': name,
      'description': description != null && description!.length > 200
          ? '${description!.substring(0, 200)}...'
          : description,
      'price': price,
      'category': category,
      'image_url': imageUrl,
    };
  }

  Product copyWith({
    String? id,
    String? userId,
    String? name,
    String? description,
    double? price,
    String? category,
    String? sku,
    String? imageUrl,
    int? stockQuantity,
    bool? isActive,
    Map<String, dynamic>? metadata,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Product(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      name: name ?? this.name,
      description: description ?? this.description,
      price: price ?? this.price,
      category: category ?? this.category,
      sku: sku ?? this.sku,
      imageUrl: imageUrl ?? this.imageUrl,
      stockQuantity: stockQuantity ?? this.stockQuantity,
      isActive: isActive ?? this.isActive,
      metadata: metadata ?? this.metadata,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  String toString() {
    return 'Product(id: $id, name: $name, price: \$$price)';
  }
}

/// Product upload response
class ProductUploadResponse {
  final int totalRows;
  final int inserted;
  final int failed;
  final List<ProductUploadError> errors;

  ProductUploadResponse({
    required this.totalRows,
    required this.inserted,
    required this.failed,
    required this.errors,
  });

  factory ProductUploadResponse.fromJson(Map<String, dynamic> json) {
    return ProductUploadResponse(
      totalRows: json['total_rows'] as int,
      inserted: json['inserted'] as int,
      failed: json['failed'] as int,
      errors: (json['errors'] as List<dynamic>?)
              ?.map(
                  (e) => ProductUploadError.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  bool get hasErrors => failed > 0;
  bool get isPartialSuccess => inserted > 0 && failed > 0;
  bool get isFullSuccess => inserted > 0 && failed == 0;
}

/// Product upload error details
class ProductUploadError {
  final int row;
  final String error;
  final Map<String, dynamic>? data;

  ProductUploadError({
    required this.row,
    required this.error,
    this.data,
  });

  factory ProductUploadError.fromJson(Map<String, dynamic> json) {
    return ProductUploadError(
      row: json['row'] as int,
      error: json['error'] as String,
      data: json['data'] as Map<String, dynamic>?,
    );
  }

  @override
  String toString() {
    return 'Row $row: $error';
  }
}
